pkg/client/introduce/example_test.go

```diff
@@ -0,0 +1,306 @@
+		Return(fakeStore(ctrl), nil).
+		AnyTimes()
+
+	// creates dispatcher
+	dispatcher := dispatcherMocks.NewMockOutbound(ctrl)
+	dispatcher.EXPECT().
+		SendToDID(gomock.Any(), gomock.Any(), gomock.Any()).
+		Do(func(msg interface{}, myDID string, theirDID string) error {
+			// converts message to bytes
+			payload, err := json.Marshal(msg)
+			if err != nil {
+				return err
+			}
+
+			// creates DIDCommMsgMap
+			didMsg, err := service.NewDIDCommMsgMap(payload)
+			if err != nil {
+				return err
+			}
+
+			// sends a message
+			transport[theirDID] <- transportMsg{msg: didMsg, myDID: myDID, theirDID: theirDID}
+
+			return nil
+		}).
+		AnyTimes()
+
+	// creates didexchange service
+	didexchangeService := serviceMocks.NewMockDIDComm(ctrl)
+	didexchangeService.EXPECT().
+		RegisterMsgEvent(gomock.Any()).
+		Return(nil)
+
+	// creates provider for the introduce
+	introduceProvider := protocolIntroduceMocks.NewMockProvider(ctrl)
+	introduceProvider.EXPECT().
+		StorageProvider().
+		Return(storageProvider)
+	introduceProvider.EXPECT().
+		OutboundDispatcher().
+		Return(dispatcher)
+	introduceProvider.EXPECT().
+		Service(didexchange.DIDExchange).
+		Return(didexchangeService, nil)
+
+	// creates introduce service
+	svc, err := introduce.New(introduceProvider)
+	if err != nil {
+		fmt.Println("introduce.New", err)
+		return nil, nil
+	}
+
+	done := make(chan struct{})
+
+	stateMsgCh := make(chan service.StateMsg)
+	if err := svc.RegisterMsgEvent(stateMsgCh); err != nil {
+		fmt.Println("svc.RegisterMsgEvent", err)
+		return nil, nil
+	}
+
+	go func() {
+		for state := range stateMsgCh {
+			if state.StateID == "done" && state.Type == service.PostState {
+				close(done)
+			}
+		}
+	}()
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case <-time.After(time.Second * 3):
+				fmt.Println("timeout")
+				close(done)
+			case inboundMsg := <-transport[agent]:
+				if inboundMsg.msg.Type() != didexchange.InvitationMsgType {
+					if _, err := svc.HandleInbound(inboundMsg.msg, inboundMsg.theirDID, inboundMsg.myDID); err != nil {
+						fmt.Println("svc.HandleInbound", err)
+						return
+					}
+				}
+
+				// mocks didexchange logic by calling directly InvitationReceived function from the introduce service
+				if err := svc.InvitationReceived(stateMsg(inboundMsg.msg)); err != nil {
+					fmt.Println("InvitationReceived", err)
+					return
+				}
+			}
+		}
+	}()
+
+	clientStorageProvider := storageMocks.NewMockProvider(ctrl)
+	clientStorageProvider.EXPECT().
+		OpenStore(introduce.Introduce).
+		Return(fakeStore(ctrl), nil).
+		AnyTimes()
+
+	introduceClientProvider := clientIntroduceMocks.NewMockProvider(ctrl)
+	introduceClientProvider.EXPECT().
+		Service(introduce.Introduce).
+		Return(svc, nil)
+	introduceClientProvider.EXPECT().
+		StorageProvider().
+		Return(clientStorageProvider)
+
+	return introduceClientProvider, done
+}
+
+// nolint: gocyclo
+func Example() {
+	wg := sync.WaitGroup{}
```

Comment:
@troyronda @fqutishat  I remember we were discussing about avoiding WaitGroups. Does that involve test file as well ?

Reply:
@rolsonquadras  test files are okay